libraries/node-core-library/src/Async.ts

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+/**
+ * Utilities for parallel asynchronous operations, to augment built-in Promise capability.
+ * @public
+ */
+export class Async {
+  /**
+   * Take an input array and map it through an asynchronous function, with a maximum number
+   * of parallel operations provided by the `parallelismLimit` parameter.
+   */
+  public static async mapLimitAsync<TEntry, TRetVal>(
+    array: TEntry[],
+    parallelismLimit: number,
+    fn: ((entry: TEntry) => Promise<TRetVal>) | ((entry: TEntry, index: number) => Promise<TRetVal>)
```

Comment:
```suggestion
    fn: (entry: TEntry, index: number) => Promise<TRetVal>
```

Someone can choose to just ignore the index parameter.

Reply:
Wouldn't this error if you pass a callback to the function that doesn't have an index param?

I guess to me this seems pretty unfriendly and veers further from the typical Array.map / Array.forEach behaviors, which I like to emulate because they are the baseline people are familiar with.

Reply:
Nope, JS is fine with that.

See https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFAFjMBzA6jVBhAQwBsiAjAiAawApgwAuRagD0YGcoAndDAGkXQATAKatEYEAFtSwzgEpEAXgB8iAG5wYguYw1bEAbwBQiU4jrUA5AEZL-a3IDcRgL5GjqHjnzEyFGizsXDwKKoYmZhAIbHBEwgB0RHAYLE6uTkA

Reply:
If that's what the `Array.map` typings say, I guess just do that.

Reply:
I think I reacted too early to an error I was seeing and made an assumption -- your example looks good, and simpler is probably better! I'll use the one without the algebra.